src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,615 @@
+
+    _addIoPort: function (type, name, value) {
+        var ioPort;
+        if (value instanceof Texture) {
+            ioPort = { type: type, name: name, valueTex: value };
+        } else if (value instanceof Vec4) {
+            ioPort = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z, valueW: value.w };
+        } else if (value instanceof Vec3) {
+            ioPort = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z };
+        } else if (value instanceof Vec2) {
+            ioPort = { type: type, name: name, valueX: value.x, valueY: value.y };
+        } else if (typeof(value) === 'number') {
+            ioPort = { type: type, name: name, valueX: value };
+        } else {
+            ioPort = { type: type, name: name }; // value is optional when adding ports
+        }
+
+        this.graphData.ioPorts.push(ioPort);
+
+        return ioPort;
+    },
+
+    addInput: function (type, name, value) {
+        return this._addIoPort(type, 'IN_' + name, value);
+    },
+
+    addOutput: function (type, name, value) {
+        return this._addIoPort(type, 'OUT_' + name, value);
+    },
+
+    addConstant: function (type, value, userId) {
+        return this._addIoPort(type, 'CONST_' + type + '_' + userId ? userId : this.graphData.ioPorts.length, value);
+    },
+
+    genCustomFuncVars: function () {
+        var functionString = this.graphData.customFuncGlsl.trim();
+
+        var head = functionString.split(")")[0];
+        var retTypeAndFuncName = head.split("(")[0];
+        var retType = retTypeAndFuncName.split(" ")[0];
+        var params = head.split("(")[1].split(",");
+
+        this.name = retTypeAndFuncName.split(" ")[1];
+
+        if (retType != "void") {
+            this.addOutput(retType, 'ret');
+        }
+
+        for (var p = 0; p < params.length; p++) {
+            var inOrOutAndTypeAndName = params[p].split(" ");
+
+            if (inOrOutAndTypeAndName[0] === "") inOrOutAndTypeAndName.shift();
+
+            if (inOrOutAndTypeAndName[0] === "out") {
+                this.addOutput(inOrOutAndTypeAndName[1], inOrOutAndTypeAndName[2]);
+            }
+            if (inOrOutAndTypeAndName[0] === "in") {
+                this.addInput(inOrOutAndTypeAndName[1], inOrOutAndTypeAndName[2]);
+            } else {
+                // unsupported parameter
+            }
+        }
+    },
+
+    addSubGraph: function (graph) {
+        var ret = this.graphData.subGraphs.length;
+        this.graphData.subGraphs.push(graph);
+        return ret;
+    },
+
+    connect: function (srcIndex, srcVarName, dstIndex, dstVarName) {
+        var connection = { srcIndex: srcIndex, srcVarName: srcVarName, dstIndex: dstIndex, dstVarName: dstVarName };
+
+        this.graphData.connections.push(connection);
+    },
+
+    _getIoPortValueString: function (ioPort) {
+        var ret;
+
+        if (ioPort.type === 'float') {
+            ret = 'float(' + ioPort.valueX + ')';

```

Comment:
do we need to make sure values for float type contain '.' - otherwise shader compiler will see them as int and often generate incompatible type error.